### PR TITLE
refactor: issue #435

### LINKED ISSUE #435: Refactor `config center` interface.
**What would you like to be added**:

**Why is this needed**:

Refactor config center interface, e.g:

`reader + watcher`: for arana runtime
`reader + writer`: for admin consule

### PR DESCRIPTION
**Pull request by chuntaojun**
**What this PR does**:

**Which issue(s) this PR fixes**:

Fixes #435 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/boot/discovery_import.go [hunk 1]
```diff
@@ -66,11 +66,16 @@ func RunImport(importConfPath, configPath string) bool {
 		tenant.Metadata = cfg.Metadata
 
 		ok := func() bool {
-			op := config.NewCenter(tenant.Name, config.GetStoreOperate())
+			op, err := config.NewCenter(tenant.Name, config.GetStoreOperate())
+			if err != nil {
+				log.Fatalf("create config_center tenant=%s failed: %+v", tenant.Name, err)
```
Comment (jjeffcaii) [inline pkg/boot/discovery_import.go:71]: The `Fatalf` will kill the current proc, use `Errorf` instead.

File: pkg/boot/discovery_import.go [hunk 2]
```diff
@@ -66,11 +66,16 @@ func RunImport(importConfPath, configPath string) bool {
...
+				return false
+			}
+
 			defer op.Close()
 
-			if err := op.Import(context.Background(), tenant); err != nil {
-				log.Fatalf("persist config to config.store failed: %+v", err)
+			if err := op.ImportAll(context.Background(), tenant); err != nil {
+				log.Fatalf("persist config tenant=%s to config.store failed: %+v", tenant.Name, err)
```
Comment (jjeffcaii) [inline pkg/boot/discovery_import.go:78]: Same as above.

File: pkg/config/config.go [hunk 3]
```diff
@@ -22,554 +22,202 @@ package config
...
 
-		tenants := make([]string, 0, 4)
-		if err := yaml.Unmarshal(val, &tenants); err != nil {
-			return err
-		}
-
-		for i := range tenants {
-			tp.tenants[tenants[i]] = struct{}{}
+		ret = &cacheConfigReader{
+			initialize: 0,
```
Comment (jjeffcaii) [inline pkg/config/config.go:111]: Remove this initialization line, because it is zero by default.

File: pkg/config/tenant.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+				tp.tenants[tenants[i]] = struct{}{}
+			}
+		}
+
+		for {
+			select {
+			case ret := <-ch:
+				consumer(ret)
+			case <-ctx.Done():
+				log.Infof("stop watch : %s", DefaultTenantsPath)
```
Comment (jjeffcaii) [inline pkg/config/tenant.go:131]: End the loop.

File: pkg/config/tenant.go [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+	if err := tp.op.Save(DefaultTenantsPath, data); err != nil {
+		return errors.Wrap(err, "create tenant name")
+	}
+
+	//need to insert the relevant configuration data under the relevant tenant
+	tenantPathInfo := NewPathInfo(name)
+	for i := range tenantPathInfo.ConfigKeyMapping {
+		if err := tp.op.Save(i, []byte("")); err != nil {
+			return errors.Wrap(err, fmt.Sprintf("create tenant resource : %s", i))
```
Comment (jjeffcaii) [inline pkg/config/tenant.go:179]: Use `errors.Wrapf(...)` instead.

**Review by jjeffcaii — COMMENTED** [review]
pls fix

File: pkg/boot/discovery.go [hunk 6]
```diff
@@ -162,7 +162,7 @@ func (fp *discovery) Import(ctx context.Context, info *config.Tenant) error {
 		return ErrorNoTenant
 	}
 
-	return op.Import(ctx, info)
+	return op.ImportAll(ctx, info)
```
Comment (dongzl) [inline pkg/boot/discovery.go:165]: why modify method name and add an `All` suffix?

File: pkg/boot/discovery.go [hunk 7]
```diff
@@ -273,7 +279,7 @@ func (fp *discovery) GetTenant(ctx context.Context, tenant string) (*config.Tena
 		return nil, ErrorNoTenant
 	}
 
-	cfg, err := op.Load(context.Background())
+	cfg, err := op.LoadAll(context.Background())
```
Comment (dongzl) [inline pkg/boot/discovery.go:282]: why

Comment (chuntaojun) [reply pkg/boot/discovery.go:282]: will fix it

Comment (chuntaojun) [reply pkg/boot/discovery.go:282]: 会单独提交一个PR再优化下这里

**Review by dongzl — APPROVED** [review]

**Review by jjeffcaii — APPROVED** [review]

**Pull request merged**